### PR TITLE
Fixes #18682 - Offloaded upload package task to dynflow executor

### PR DESCRIPTION
**Pull request by ShimShtein**
It's a first try, let's see that all the tests are passing

### CODE REVIEW

File: app/controllers/katello/api/rhsm/candlepin_dynflow_proxy_controller.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
+module Katello
+  class Api::Rhsm::CandlepinDynflowProxyController < Api::V2::ApiController
```
Comment (jlsherrill) [inline app/controllers/katello/api/rhsm/candlepin_dynflow_proxy_controller.rb]: Curious, why did you move this to another controller?

Comment (ShimShtein) [inline app/controllers/katello/api/rhsm/candlepin_dynflow_proxy_controller.rb]: Part preparing for the future, part for the sake of clarity.

In rails 5, parameter parsing is done by base controller, so we will need a different inheritance for upload_packages action.
In rails 4, as I found out later, the parsing is done by rack middleware, so I don't have to inherit `Metal` directly.

Comment (ShimShtein) [inline app/controllers/katello/api/rhsm/candlepin_dynflow_proxy_controller.rb]: I expect this action getting a completely different handling, since it won't have the payload serialized.

File: lib/katello/params_parser_wrapper.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
+module Katello
+  class ParamsParserWrapper < ActionDispatch::ParamsParser
```
Comment (ehelms) [inline lib/katello/params_parser_wrapper.rb]: Can you give more details on what this is doing?

**Comment by ehelms** [comment]
Going to summarize what I think this is doing so I know if I got the idea right. 

Current Problem:
Package profiles come in as a large JSON string that is parsed into a hash and then passed to the task. Given the size of the hash being passed, both the planning phase takes longer and due to the size of the input dynflow increases the memory footprint it has.

Solution Presented Here:
Instead of letting Rails parse the incoming params JSON to a hash and pass that to the task. Take the raw JSON string, pass that to the task and let the task parse the JSON.

Let me know if I have missed any nuances. Based on my "understanding", I have the following questions:

 1) This seems like a generic problem with Foreman Tasks or Dynflow where we have to use this pattern every time there is largish input. I would love to see this fixed systematically if possible to prevent this edge cases and having to work around them.
 2) If the solution is just to pass the raw string in and have it parsed, that feels like it should be a 2-3 line code change in the controller and the original action.

**Comment by ShimShtein** [comment]
@ehelms, You got the basics, I'll clarify more:

Process before the PR: 
* `upload_package_profile` receives a huge JSON, parses it into hash and invokes in-process `UploadPackageProfile` action's `plan` phase.
* The plan action calls to `Pulp::Consumer` with the hash, syncs the state of host model by calling `import_package_profile` and plans `GenerateApplicability` task.

Problems with this state: 
1. The JSON is parsed inside passenger process boundaries, causing instantiation of ~10K objects, which puts a lot of strain on GC causing those objects to get stuck in memory for long times.
2. All actions are performed inside UI request-response cycle (synchronously), taking resources from Passenger for unnecessary long time.

What this PR is suggesting:
This PR makes the UI process responsible only for passing the unparsed string into dynflow DB. All the JSON parsing and original planning are done by the new `UploadPackageProfileWrapper` task as part of dynflow executor process.
Why is it better:
The basic concept is moving asynchronous handling from our UI process to dynflow executor, which could be tuned better for this kind of tasks. The tuning includes ruby GC parameter tuning, dynflow tasks throttling (concurrency model) and recycling policy.

>  I would love to see this fixed systematically if possible to prevent this edge cases and having to work around them.

You are right, once we see the improvement and a second candidate for it, I will gladly extract common parts into framework.

> If the solution is just to pass the raw string in and have it parsed, that feels like it should be a 2-3 line code change in the controller and the original action.

I want to eliminate JSON parsing altogether. The problem with this is that Rails parses all incoming parameters automatically by [Rack middleware](https://github.com/rails/rails/blob/4-1-stable/actionpack/lib/action_dispatch/middleware/params_parser.rb#L23). Since I want to eliminate this, I had to "cancel" this middleware for our action (Actually this answers your other question :) ). Then I had to wrap the Dynflow action in order to make the action parse the parameters instead of the controller. Yes, I could change the original action to include all the code from the wrapper, but I preferred to make these changes in a separate file, so later extraction into framework could go easier. Plus the code is more clear this way - each action has its own responsibility.

File: lib/katello/params_parser_wrapper.rb (hunk 2 above)
Comment (ShimShtein) [inline lib/katello/params_parser_wrapper.rb]: Answered in the lower part of [this](https://github.com/Katello/katello/pull/6640#issuecomment-284220686) comment.

**Comment by jlsherrill** [comment]
[test]

File: app/controllers/katello/api/rhsm/candlepin_dynflow_proxy_controller.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    def upload_package_profile
+      User.as_anonymous_admin do
+        async_task(::Actions::Katello::Host::UploadPackageProfileWrapper, @host, request.body.string)
+      end
+      [
+        200,
+        { "Content-Type" => "application/json" },
+        Resources::Candlepin::Consumer.get(@host.subscription_facet.uuid)
+      ]
+    end
```
Comment (jlsherrill) [inline app/controllers/katello/api/rhsm/candlepin_dynflow_proxy_controller.rb]: can you think of anyway to write a test to test that would test that the body isn't parsed?  the code handling that seems a bit fragile so it would be nice if we could test that.  Not a pure blocker to merging though.

**Comment by jlsherrill** [comment]
Do you have any memory stats from before and after this change for both passenger and foreman-tasks?

File: app/controllers/katello/api/rhsm/candlepin_dynflow_proxy_controller.rb (hunk 3 above)
Comment (ShimShtein) [inline app/controllers/katello/api/rhsm/candlepin_dynflow_proxy_controller.rb]: It depends if the actions are executed during the test. If they are not executed, then we could count total number of allocations delta (which should go down significally). If the actions are performed, then we have a problem - I'm just moving the allocations into the action, so they will still be counted.

Comment (jlsherrill) [inline app/controllers/katello/api/rhsm/candlepin_dynflow_proxy_controller.rb]: you can easily stub the controller .async_task action, we do that in a number of places such as: 

https://github.com/Katello/katello/blob/master/test/controllers/api/v2/content_view_versions_controller_test.rb#L123

so you could test before and after the action with the test stub and count the allocation delta, that would be a good test!

**Comment by ShimShtein** [comment]
- [x] Added a test to see that there is no JSON deserialization during the action.
- [x] Fixed paths to include both `/packages` and `/profile`
- [x] Fixed rendering error.

**Comment by ShimShtein** [comment]
Fixed "Unauthorized" error

**Comment by jlsherrill** [comment]
@ShimShtein in general this looks good. I think the wrapper action is unnecessary though and adds junk for the user to wade through.  When updating a package on 1000 clients for example, they'd have 2000 tasks created instead of 1000.  Could we just move the parsing and actual work to the run phase of the UploadPackageProfile?

**Comment by jlsherrill** [comment]
and i realize that you said we'd like to build this into foreman-tasks/dynflow but that may be some time and I'm not a huge fan of having this wrapper action in the meantime.

**Comment by ShimShtein** [comment]
- [x]  Removed wrapper task.
- [x] Removed commented out code.

File: app/controllers/katello/api/rhsm/candlepin_dynflow_proxy_controller.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    skip_before_action :authorize, :only => [:upload_package_profile]
+    before_action :find_host, :only => [:upload_package_profile]
+    before_action :authorize_client_or_user, :only => [:upload_package_profile]
+
+    #api :PUT, "/consumers/:id/packages", N_("Update installed packages")
+    #api :PUT, "/consumers/:id/profile", N_("Update installed packages")
+    #param :id, String, :desc => N_("UUID of the consumer"), :required => true
+    def upload_package_profile
+      User.as_anonymous_admin do
+        async_task(::Actions::Katello::Host::UploadPackageProfile, @host, request.body.string)
```
Comment (ehelms) [inline app/controllers/katello/api/rhsm/candlepin_dynflow_proxy_controller.rb]: This needs to be `request.body.read` or errors like the following can occur:

```
NoMethodError: undefined method `string' for #<PhusionPassenger::Utils::TeeInput:0x007f882fcba5e8>
```

**Comment by ShimShtein** [comment]
Fixed tests and controller

File: lib/katello/engine.rb [hunk 5]
```diff
@@ -2,6 +2,15 @@ module Katello
   class Engine < ::Rails::Engine
     isolate_namespace Katello
 
+    initializer 'katello.selective_params_parser', :before => :load_config_initializers do |app|
+      require 'katello/params_parser_wrapper'
+      app.config.app_middleware.swap(
+      ActionDispatch::ParamsParser,
+      Katello::ParamsParserWrapper,
+      ->(env) { env['PATH_INFO'] =~ /consumers/ && env['PATH_INFO'] =~ /profile|packages/ }
+      )
```
Comment (ehelms) [inline lib/katello/engine.rb]: This didn't work for me. Instead I had to change it to:

```
initializer 'katello.selective_params_parser', :before => :build_middleware_stack do |app|
      require 'katello/params_parser_wrapper'
      app.middleware.swap(
        ActionDispatch::ParamsParser,
        Katello::ParamsParserWrapper,
        ->(env) { env['PATH_INFO'] =~ /consumers/ && env['PATH_INFO'] =~ /profile|packages/ }
      )
end
```

**Comment by ShimShtein** [comment]
@ehelms Changed the initializer to the suggested.

File: lib/katello/engine.rb [hunk 6]
```diff
@@ -2,6 +2,15 @@ module Katello
   class Engine < ::Rails::Engine
     isolate_namespace Katello
 
+    initializer 'katello.selective_params_parser', :before => :build_middleware_stack do |app|
+      require 'katello/params_parser_wrapper'
+      app.middleware.swap(
+      ActionDispatch::ParamsParser,
+      Katello::ParamsParserWrapper,
+      ->(env) { env['PATH_INFO'] =~ /consumers/ && env['PATH_INFO'] =~ /profile|packages/ }
```
Comment (ehelms) [inline lib/katello/engine.rb]: Nitpick but these params to the method should be indented.

**Comment by ehelms** [comment]
[test]

**Comment by ShimShtein** [comment]
Fixed indentation.

**Comment by ehelms** [comment]
I'm not sure why this test is failing here when all you did was move the test. The rest of the code looks good, and I'm going to test the use of `raw_post` to ensure we don't hit another error as we did with the `string` method.

**Comment by ShimShtein** [comment]
[test]

**Comment by ShimShtein** [comment]
Indeed it looks weird...
From what I see, the error comes from test_helper's `get_organization` method, and it complains that it tries to remove a location. Could it be some kind of race condition? Anyway I don't understand what this method has to do with removing locations... (the actual error is that a location cannot be removed, since it has hasts connected to it).

**Comment by ehelms** [comment]
@ShimShtein It's almost as if when it saves the user, maybe its removing locations from the user and the user isn't in a clean state?

**Comment by ShimShtein** [comment]
Still don't know what caused the error, but I don't need `@organization` variable at all. Removed the line altogether.

From the investigation it looks like the error message is not accurate - when trying to save the organization, it looks for attached locations and finds one but the list is supposed to be empty.

**Comment by ehelms** [comment]
@jlsherrill do you want to have a final go at reviewing this?

**Comment by jlsherrill** [comment]
ACK, sorry for the delay this slipped under my rader.  Thanks @ShimShtein !
